apiserver/facades/client/action/action.go

```diff
@@ -184,11 +213,36 @@ func (a *ActionAPI) Enqueue(arg params.Actions) (params.ActionResults, error) {
 		return params.ActionResults{}, errors.Trace(err)
 	}
 
+	var leaders map[string]string
+	getLeader := func(appName string) (string, error) {
```

Comment:
why have this be a closure rather than just a method on ActionAPI? Is it because you wanted to cache the lookups? Do we think there will be a lot of lookups that it is worth doing?
We could always make it an optional cache to pass in to the method. (you could always pass nil, though we would have to check before setting in in the cache)

Reply:
I imagine it will be rare indeed to be resolving more than one leader in a call.

I guess the intent here is that there are _zero_ calls to `ApplicationLeaders` if the are no leaders to resolve, and the closure allows you to keep leaders as method-local variable.